### DISCUSSION TITLE
blog/react-query-fa-qs

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#81] [Announcements]
# blog/react-query-fa-qs

Answering the top 3 most frequently asked React Query questions

https://tkdodo.eu/blog/react-query-fa-qs#inject-comments

**Comment by Moshyfawn** [comment]
I didn't know I needed [the "Why `useQueryClient`" info](https://tkdodo.eu/blog/react-query-fa-qs#why-should-i-usequeryclient), but it was nice to find answers to the questions you have in the back of your mind while working with a tool, not having time to dig into details of it's implementation or the reasoning behind it.

Awesome work, as always!

**Comment by dcheng1290** [comment]
Nice blog!

**Comment by eminvergil** [comment]
It was nice to read how you give a detail for every scenerio and possible different answers. Good article, thx.

**Comment by nathanelcorpuz** [comment]
I am using REST and have my own error handling implementation from the backend. The "correct-logging-example" section helped a lot after 1 whole day of debugging. Thank you.

**Reply by TkDodo** [reply]
You're welcome. I'm always happy to hear that my blog is helpful for other developers 😊

**Comment by lyan-ap** [comment]
may I met the callback hell again...

**Comment by william-hotmart** [comment]
There is some example of best way to organize a project that uses Tanstack Query?

Great job always =)

**Comment by Gitter22** [comment]
I have few queries:
1. I have list data with multiple drop down filters. Some filters are independent, some are cascading(meaning dependent on some other filters selection) Does it make sense to use useQuery for fetching data for filters apart from the main listing api?
2. When I have multiple useQuery calls on the page(like the above example where there are api calls for dropdowns as well as main list), I want to alert user if any of the api call fails via a toaster alert with a dedicated message. What is a better way to handle errors in this case, if using onError callback is soon going to be deprecated. Do i write multiple useEffects?
3. If the initial fetch fails, is there a way to keep the placeholder data(note, i cant use initialData as it messes up my stale time option) instead of data becoming undefined. I want placeholder data to be empty array, but as soon as the fetch fails, it becomes undefined which breaks my UI. No matter how i structure my query function, the data always becomes undefined and I have to use conditional chaining.

**Comment by gboycdw** [comment]
Thank you. I was able to easily implement an infinite scroll page using useInfiniteQuery.

However, when the staleTime of the data fetched with useInfiniteQuery expires and I visit a different page and then come back, multiple query data for each cursor that has been created ends up being refetched.

Is there a way to debounce or artificially slow down the refetching process?

The data used on this page is very heavy, and I have no choice but to limit the rate of consecutive API requests from the server.

For example, if I have 5 cursors and 6 data sets on a page, and I visit a different page and come back, I end up with 6 refetching operations, and I would like to intentionally slow down the speed of those 6 API calls. Is there a way to do that?

**Comment by lili21** [comment]
what will happen if I call  useQuery multiple time with same key but different options?

```js

useQuery(['todos'], queryFn, { staleTime: 0 })

useQuery(['todos'], queryFn, { staleTime: 5 })
```

**Reply by TkDodo** [reply]
depends on the option. If it's something that is observer specific, like `refetchInterval`, each one will have their own. For `staleTime`, the lowest one will be taken, for `gcTime` the maximum.

**Comment by lili21** [comment]
can I using `startTransiton/useTransition` with suspense to handle the loading state?

**Reply by lili21** [reply]
I can trying to change the queryKey inside `startTranstion`, but the query still fallback to suspense.

**Reply by lili21** [reply]
I found the doc and issue. I should enable startTransition in react-router.

**Comment by rikotacards** [comment]
I've had an issue, where the solution seems a bit over-engineered and complicated. We have an endpoint, lets call it offset, we query this offset value ever 2.5 seconds. The value is monotonically increasing. 

We have another endpoint, say getTransactions, where we have to pass the offset value in. Lets say the query key is just 'getTransactions'
At time 0, we do getTransactions(0), 2.5 seconds later, getTransactions(2) ... getTransactions(3). 

However, getTransactions might take 5 seconds to respond. So i'd we might be trying to call it with getTransactions(3), but actually we'd want to hold off, until getTransactions(2) is complete. 

Now, at first, we had the offset as part of the query key, because on the dev environment, that value changes slowly. So the endpoint offset might return the same value over a period of time. 

I've had to remove the offset value from the query key, so before ['getTransactions', offset],  --> ['getTransactions'], However, now that the offset is outside of the query key, how do we trigger a fetch for every poll? 

I've used AI , and it has suggested many useEffect solutions, that involve `refetch()`, bur refetch bypasses the dedupe feature. So I have have two components that use the same hook, I get 2 network requests... Running out of ideas!

**Reply by TkDodo** [reply]
> but actually we'd want to hold off, until getTransactions(2) is complete.

I think you’d still want `['getTransactions', offset]` in the key, but implement some sort of queueing on the actual API layer.

One way to do this is by using `@tantstack/pacer`: https://tanstack.com/pacer/latest/docs/guides/queuing

Adding a `queue` with a size of one around this specific request should do the trick.
